### PR TITLE
Avoid options field allocation in UnnamedOptionsManager

### PR DESCRIPTION
**Pull request by sungam3r**
Follow-up to #49883

### CODE REVIEW

File: src/libraries/Microsoft.Extensions.Options/src/UnnamedOptionsManager.cs [hunk 1]
```diff
@@ -10,24 +10,23 @@ internal sealed class UnnamedOptionsManager<[DynamicallyAccessedMembers(Options.
...
-                if (_value is TOptions value)
+                if (_factoryOrValue is TOptions value)
                 {
                     return value;
                 }
 
                 lock (_syncObj ?? Interlocked.CompareExchange(ref _syncObj, new object(), null) ?? _syncObj)
                 {
-                    return _value ??= _factory.Create(Options.DefaultName);
+                    return _factoryOrValue ??= ((IOptionsFactory<TOptions>)_factoryOrValue).Create(Options.DefaultName);
```
Comment (sungam3r) [inline src/libraries/Microsoft.Extensions.Options/src/UnnamedOptionsManager.cs:29]:
```suggestion
                    return (TOptions)(_factoryOrValue ??= ((IOptionsFactory<TOptions>)_factoryOrValue).Create(Options.DefaultName));
```

File: src/libraries/Microsoft.Extensions.Options/src/UnnamedOptionsManager.cs [hunk 2]
```diff
@@ -10,24 +10,23 @@ internal sealed class UnnamedOptionsManager<[DynamicallyAccessedMembers(Options.
...
-                if (_value is TOptions value)
+                if (_factoryOrValue is TOptions value)
                 {
                     return value;
                 }
 
                 lock (_syncObj ?? Interlocked.CompareExchange(ref _syncObj, new object(), null) ?? _syncObj)
                 {
-                    return _value ??= _factory.Create(Options.DefaultName);
+                    return (TOptions)(_factoryOrValue ??= ((IOptionsFactory<TOptions>)_factoryOrValue).Create(Options.DefaultName));
```
Comment (sungam3r) [inline src/libraries/Microsoft.Extensions.Options/src/UnnamedOptionsManager.cs:29]:
```suggestion
                    return (TOptions)(_factoryOrValue ??= ((IOptionsFactory<TOptions>)_factoryOrValue!).Create(Options.DefaultName));
```

**Comment by tarekgh** [comment]
@sungam3r can you paste the perf number showing how much this is going to save. I am afraid this is not worth the optimization. 

CC @stephentoub as mentioned this change is a follow up from https://github.com/dotnet/runtime/pull/49883

**Comment by sungam3r** [comment]
>  paste the perf number showing how much this is going to save

Should I use some existing perf tooling from that repo?

**Comment by sungam3r** [comment]
You are right. Benchmark shows performance decrease.

**Pull request closed without merging**